protocols/relay/src/priv_client.rs

```diff
@@ -126,6 +130,14 @@ impl Behaviour {
                     unreachable!("`on_connection_closed` for unconnected peer.")
                 }
             };
+            if let Some(addr) = self.relay_connection_addr.remove(&connection_id) {
+                let addr = addr
+                    .with(Protocol::P2pCircuit)
+                    .with(Protocol::P2p(self.local_peer_id));
+
+                self.queued_actions
+                    .push_back(ToSwarm::ExternalAddrExpired(addr));
+            }
```

Comment:
If the reservation fails and then we close the connection, this will still emit an event, right?

It would likely be more resilient to have a 2nd data structure, like a `HashSet` that tracks our relayed addresses and use that to ensure we emit the correct event.

Like:

- When `insert`ed into the hashset and it already exists, don't emit an event
- When the connection closes and we have a relay address for this connection, remove it. If `remove` returns `true`, emit the event.

That could also solve the renewal, not sure what is "cleaner" to implement there. If `renewal` is `true`, the address should really already be in the `HashSet`. Perhaps we can `debug_assert` that?